File: app/src/test/java/com/hana053/micropost/presentation/core/services/NavigatorTest.java
package com.hana053.micropost.presentation.core.services;

import android.app.Activity;
import android.content.Intent;

import com.hana053.micropost.domain.User;
import com.hana053.micropost.presentation.login.LoginActivity;
import com.hana053.micropost.presentation.main.MainActivity;
import com.hana053.micropost.presentation.relateduserlist.followerlist.FollowerListActivity;
import com.hana053.micropost.presentation.relateduserlist.followinglist.FollowingListActivity;
import com.hana053.micropost.presentation.signup.SignupActivity;
import com.hana053.micropost.presentation.usershow.UserShowActivity;
import com.hana053.micropost.testing.RobolectricBaseTest;

import org.junit.Before;
import org.junit.Test;
import org.parceler.Parcels;
import org.robolectric.Robolectric;
import org.robolectric.Shadows;
import org.robolectric.shadows.ShadowActivity;

import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

public class NavigatorTest extends RobolectricBaseTest {

    private Activity activity;
    private Navigator navigator;

    @Before
    public void setup() {
        activity = Robolectric.setupActivity(Activity.class);
        navigator = new Navigator(activity);
    }

    @Test
    public void shouldNavigateToUserShow() {
        final User user = new User(1, "test", "test@test.com", "");
        navigator.navigateToUserShow(user);
        final ShadowActivity shadowActivity = Shadows.shadowOf(activity);
        final Intent intent = shadowActivity.getNextStartedActivity();
        assertThat(intent.getComponent().getClassName(), is(UserShowActivity.class.getName()));
        assertThat(Parcels.unwrap(intent.getExtras().getParcelable(UserShowActivity.KEY_USER)), equalTo(user));
    }

    @Test
    public void shouldNavigateToMain() {
        navigator.navigateToMain();
        final ShadowActivity shadowActivity = Shadows.shadowOf(activity);
        final Intent intent = shadowActivity.getNextStartedActivity();
        assertThat(intent.getComponent().getClassName(), is(MainActivity.class.getName()));
    }

    @Test
    public void shouldNavigateToSignup() {
        navigator.navigateToSignup();
        final ShadowActivity shadow = Shadows.shadowOf(activity);
        final Intent intent = shadow.getNextStartedActivity();
        assertThat(intent.getComponent().getClassName(), is(SignupActivity.class.getName()));
    }

    @Test
    public void shouldNavigateToLogin() {
        navigator.navigateToLogin();
        final ShadowActivity shadow = Shadows.shadowOf(activity);
        final Intent intent = shadow.getNextStartedActivity();
        assertThat(intent.getComponent().getClassName(), is(LoginActivity.class.getName()));
    }

    @Test
    public void shouldNavigateToFollowings() {
        navigator.navigateToFollowings(1);
        final ShadowActivity shadowActivity = Shadows.shadowOf(activity);
        final Intent intent = shadowActivity.getNextStartedActivity();
        assertThat(intent.getComponent().getClassName(), is(FollowingListActivity.class.getName()));
        assertThat(intent.getExtras().getLong(FollowingListActivity.KEY_USER_ID), is(1L));
    }

    @Test
    public void shouldNavigateToFollowers() {
        navigator.navigateToFollowers(1);
        final ShadowActivity shadowActivity = Shadows.shadowOf(activity);
        final Intent intent = shadowActivity.getNextStartedActivity();
        assertThat(intent.getComponent().getClassName(), is(FollowerListActivity.class.getName()));
        assertThat(intent.getExtras().getLong(FollowingListActivity.KEY_USER_ID), is(1L));
    }

}

File: app/src/test/java/com/hana053/micropost/service/NavigatorTest.kt
package com.hana053.micropost.service

import android.app.Activity
import android.content.Intent
import com.hana053.micropost.pages.login.LoginActivity
import com.hana053.micropost.pages.main.MainActivity
import com.hana053.micropost.pages.main.MainActivity.Companion.REQUEST_POST
import com.hana053.micropost.pages.micropostnew.MicropostNewActivity
import com.hana053.micropost.pages.relateduserlist.RelatedUserListActivity
import com.hana053.micropost.pages.signup.SignupActivity
import com.hana053.micropost.pages.usershow.UserShowActivity
import com.hana053.micropost.testing.RobolectricBaseTest
import com.hana053.micropost.testing.TestUser
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before
import org.junit.Test
import org.robolectric.Robolectric
import org.robolectric.Shadows


class NavigatorTest : RobolectricBaseTest() {

    lateinit private var activity: TestActivity
    lateinit private var navigator: Navigator

    @Before
    fun setup() {
        activity = Robolectric.setupActivity(TestActivity::class.java)
        navigator = NavigatorImpl(activity)
    }

    @Test
    fun shouldNavigateToUserShow() {
        val user = TestUser

        navigator.navigateToUserShow(user.id)

        val shadow = Shadows.shadowOf(activity)
        val intent = shadow.nextStartedActivity

        assertThat(intent.component.className, `is`(UserShowActivity::class.java.name))
        assertThat(intent.extras.getLong(UserShowActivity.KEY_USER_ID), `is`(1L))
    }

    @Test
    fun shouldNavigateToMain() {
        navigator.navigateToMain()

        val shadow = Shadows.shadowOf(activity)
        val intent = shadow.nextStartedActivity

        assertThat(intent.component.className, `is`(MainActivity::class.java.name))
    }

    @Test
    fun shouldNavigateToSignup() {
        navigator.navigateToSignup()

        val shadow = Shadows.shadowOf(activity)
        val intent = shadow.nextStartedActivity

        assertThat(intent.component.className, `is`(SignupActivity::class.java.name))
    }

    @Test
    fun shouldNavigateToLogin() {
        navigator.navigateToLogin()

        val shadow = Shadows.shadowOf(activity)
        val intent = shadow.nextStartedActivity

        assertThat(intent.component.className, `is`(LoginActivity::class.java.name))
    }

    @Test
    fun shouldNavigateToFollowings() {
        navigator.navigateToFollowingList(1)

        val shadow = Shadows.shadowOf(activity)
        val intent = shadow.nextStartedActivity

        assertThat(intent.component.className, `is`(RelatedUserListActivity::class.java.name))
        assertThat(intent.extras.getLong(RelatedUserListActivity.KEY_USER_ID), `is`(1L))
        assertThat(
            intent.extras.getSerializable(RelatedUserListActivity.KEY_LIST_TYPE) as RelatedUserListActivity.ListType,
            `is`(RelatedUserListActivity.ListType.FOLLOWING)
        )
    }

    @Test
    fun shouldNavigateToFollowers() {
        navigator.navigateToFollowerList(1)

        val shadow = Shadows.shadowOf(activity)
        val intent = shadow.nextStartedActivity

        assertThat(intent.component.className, `is`(RelatedUserListActivity::class.java.name))
        assertThat(intent.extras.getLong(RelatedUserListActivity.KEY_USER_ID), `is`(1L))
        assertThat(
            intent.extras.getSerializable(RelatedUserListActivity.KEY_LIST_TYPE) as RelatedUserListActivity.ListType,
            `is`(RelatedUserListActivity.ListType.FOLLOWER)
        )
    }

    @Test
    fun shouldNavigateToMicropostNew() {
        navigator.navigateToMicropostNew()

        val shadow = Shadows.shadowOf(activity)
        val intent = shadow.nextStartedActivity

        assertThat(intent.component.className, `is`(MicropostNewActivity::class.java.name))

        shadow.receiveResult(
            Intent(activity, MicropostNewActivity::class.java),
            Activity.RESULT_OK,
            Intent()
        )
        assertThat(activity.postReceived, `is`(true))
    }

    class TestActivity : Activity() {
        var postReceived = false

        override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
            super.onActivityResult(requestCode, resultCode, data)
            if (requestCode == REQUEST_POST) postReceived = true
        }
    }

}